drivers/volume/portworx/portworx.go

```diff
@@ -2952,6 +2939,107 @@ func (d *portworx) upgradePortworxDaemonset(specGenUrl string) error {
 	return nil
 }
 
+// upgradePortworxStorageCluster upgrades PX StorageCluster based on the given Spec Generator URL
+func (d *portworx) upgradePortworxStorageCluster(specGenUrl string) error {
+	log.InfoD("Upgrading Portworx StorageCluster")
+
+	// Get k8s version
+	k8sVersion, err := k8sCore.GetVersion()
+	if err != nil {
+		return err
+	}
+
+	// Get images from version URL
+	imageList, err := optest.GetImagesFromVersionURL(specGenUrl, k8sVersion.String())
+	if err != nil {
+		return fmt.Errorf("failed to get image list from version URL [%s], Err: %v", specGenUrl, err)
+	}
+
+	// Get PX StorageCluster
+	cluster, err := d.GetDriver()
+	if err != nil {
+		return err
+	}
+
+	// Update PX Image and Env vars if needed
+	updateParamFunc := func(cluster *v1.StorageCluster) *v1.StorageCluster {
+		// Set oci-mon version
+		cluster.Spec.Image = imageList["version"]
+		var envVars []corev1.EnvVar
+		// Add Release Manifest URL incase of edge spec
+		envVars, err := configurePxReleaseManifestEnvVars(cluster.Spec.Env, specGenUrl)
+		if err != nil {
+			return nil
+		}
+		cluster.Spec.Env = envVars
+		return cluster
+	}
+
+	// Update and validate PX StorageCluster
+	if _, err := d.updateAndValidateStorageCluster(cluster, updateParamFunc, specGenUrl, false); err != nil {
+		return err
+	}
+
+	log.Infof("Successfully upgraded Portworx StorageCluster [%s]", cluster.Name)
+	return nil
+}
+
+// upgradePortworxOperator upgrades PX Operator based on the given Spec Generator URL
+func (d *portworx) upgradePortworxOperator(specGenUrl string) error {
+	log.InfoD("Upgrading Portworx Operator")
+
+	pxOperatorSpecFileName := "/px-operator.yaml"
+
+	// Get k8s version
+	kubeVersion, err := d.schedOps.GetKubernetesVersion()
+	if err != nil {
+		return err
+	}
+
+	// Getting PX Operator spec
+	u, err := url.Parse(specGenUrl)
+	if err != nil {
+		fmt.Printf("failed to parse URL [%s], Err: %v", specGenUrl, err)
```

Comment:
log?

Reply:
Good catch, fix it in another place I copied this func from